crypto/zk/address.go

```diff
@@ -0,0 +1,122 @@
+package zk
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/iden3/go-iden3-crypto/babyjub"
+	"github.com/iden3/go-iden3-crypto/poseidon"
+	"go.vocdoni.io/dvote/crypto/ethereum"
+	"go.vocdoni.io/dvote/tree/arbo"
+	"go.vocdoni.io/dvote/types"
+	"go.vocdoni.io/dvote/util"
+)
+
+// defaultZkAddrLen contains the default size of a ZkAddress in bytes.
+const defaultZkAddrLen = 20
+
+// ZkAddress struct allow to create and encoding properly the zk-snark
+// compatible address for the vochain. This address is calculated from a
+// BabyJubJub key pair, based on a seed. The address has 20 bytes of size and
+// it is the truncated version of the poseidon hash of the BabyJubJub publicKey.
+type ZkAddress struct {
+	// Privkey contains the big.Int version of the BabyJubJub private key
+	PrivKey *big.Int
+	// PubKey contains the big.Int poseidon hash of the BabyJubJub public key
+	// (x and y coordinates of a point of the curve)
+	PubKey *big.Int
+	// addr contains the public key big.Int reduced to 20 bytes
+	// (defaultZkAddrLen). It is a private attribute because the correct formats
+	// are calulated throw other struct methods such as ZkAddress.String() or
+	// ZkAddress.Bytes().
+	addr *big.Int
+}
+
+// AddressFromBytes functions returns a new ZkAddress based on the seed
+// provided. Using it, a BabyJubJub key pair is calculated and based on it. The
+// ZkAddress components are:
+//   - ZkAddress.PrivKey: The BabyJubJub private key big.Int value, following
+//     the EdDSA standard, and using blake-512 hash.
+//   - ZkAddress.PublicKey: The poseidon hash of the BabyJubJub public key
+//     components (X and Y coordinates as big.Int).
+//   - ZkAddress.addr: The truncated to 20 bytes (defaultZkAddrLen) little
+//     endian version of the ZkAddress.PublicKey.
+func AddressFromBytes(seed []byte) (*ZkAddress, error) {
+	if len(seed) < 16 {
+		return nil, fmt.Errorf("the seed provided does not have 32 bytes at least")
```

Comment:
I try to answer yo you @elboletaire, but I am not an expert. I discovered `fmt.Errorf` when I started to work on Vocdoni so I don't know all the advantages or disadvantages between both ways, buy from my experience, creating errors with `fmt.Errorf` allows to wrap an error (like `errors.Wrap`) but also to include a formatted string (using templates). For example, if you want to wrap a received error, and also add more context with some variables:

- Using `errors.Wrap`:
```go
input := "some parameter"
err := someFuncThatFails(input)
if err != nil {
    errMsg := fmt.Sprintf("something fails with input '%s'", input)
    return errors.Wrap(err, errMsg)
}
```

- Using `errors.New`:
```go
input := "some parameter"
err := someFuncThatFails(input)
if err != nil {
    errMsg := fmt.Sprintf("something fails with input '%s': %w", input, err)
    return errors.New(errMsg)
}
```

- Using `fmt.Errorf`:
```go
input := "some parameter"
err := someFuncThatFails(input)
if err != nil {
    return fmt.Errorf("something fails with input '%s': %w", input, err)
}
```

Reply:
@altergui I updated the restriction to the correct number of bytes (32)

Reply:
@elboletaire long ago we decided to use only `fmt.Errorf()` for convention. Since it allows formatting, strings is much more convenient. For those cases where the string does not need to be formatted, it just works.